fix(daemon): handle termination signal when no coordinators found instead of infinite loop

wiseaidev opened the pull request:
Hello everyone 😊,

I recently had a great conversation with [Han Su](https://github.com/suhan1996) and [Yong He](https://www.linkedin.com/in/yong-he-1334902) about this framework, which quickly caught my interest due to my background in edge computing and the Zenoh protocol.

Mr. He mentioned that the team is looking for core maintainers, and I expressed my interest. Over the past few days, I've put together a **proposal to improve the library**, which you can find [here](https://github.com/wiseaidev/dora-proposal).

To give a practical example of potential improvements, I'm submitting this **sample PR** that addresses one of the areas outlined in the proposal. There are also several related ideas listed in open issues.

I'd love to know your thoughts on both the proposal and this initial PR.

Best,
_Mahmoud_

haixuanTao commented:
Hey Mahmoud!

Thanks for this PR. Could it be possible to reuse the ctrl-c handler at: https://github.com/dora-rs/dora/blob/f141817bbf5d1a6de6cc842535c2ad82ccd8cff2/binaries/daemon/src/lib.rs#L104C1-L105C1 instead of introducing a new one in order to keep track of those listener.

wiseaidev commented:
> Hey Mahmoud!
> 
> Thanks for this PR. Could it be possible to reuse the ctrl-c handler at: https://github.com/dora-rs/dora/blob/f141817bbf5d1a6de6cc842535c2ad82ccd8cff2/binaries/daemon/src/lib.rs#L104C1-L105C1 instead of introducing a new one in order to keep track of those listener.

Makes sense! Fixed!

wiseaidev commented:
> > > Hey Mahmoud!
> > > Thanks for this PR. Could it be possible to reuse the ctrl-c handler at: https://github.com/dora-rs/dora/blob/f141817bbf5d1a6de6cc842535c2ad82ccd8cff2/binaries/daemon/src/lib.rs#L104C1-L105C1 instead of introducing a new one in order to keep track of those listener.
> > 
> > 
> > Makes sense! Fixed it!
> 
> Do you mind checking the CI?

It turns out that registering CtrlC events twice isn't possible. Luckily tho, the integration tests caught this issue. To address this, I moved the registration of the CtrlC event handler to the coordinator init phase that is set up once.

haixuanTao reviewed (changes requested):
I'm sorry but you can't just delete the previous Ctrl-C handler just to push new change...
Making ctrl-c not workable for the rest of the process...

phil-opp commented:
I opened https://github.com/dora-rs/dora/pull/758 to check for ctrl-c events during the initial connection setup. Instead of checking it inside the `register` function, I added a `select` around all event stream setup functions directly in `Daemon::run`.

haixuanTao commented:
Superseded by #758
